consensus/src/block_storage/sync_manager.rs

```diff
@@ -308,10 +324,45 @@ impl BlockStore {
             highest_quorum_cert,
         );
 
-        // we fetch the blocks from
-        let num_blocks = highest_quorum_cert.certified_block().round()
-            - highest_commit_cert.ledger_info().ledger_info().round()
-            + 1;
+        // TODO: we should not retrieve genesis, double check this is the right way to do that
+        // If execution pool is enabled, use round based block retrieval, else use target block id
+        let (target_block_retrieval_payload, num_blocks) = match window_size {
+            None => {
+                let num_blocks = highest_quorum_cert.certified_block().round()
+                    - highest_commit_cert.ledger_info().ledger_info().round()
+                    + 1;
+                let target_block_id = highest_commit_cert.commit_info().id();
+                info!(
+                    "fast_forward_sync with NO window_size: {:?}, target_block_id: {}, num_blocks: {}",
+                    window_size, target_block_id, num_blocks
+                );
+                (
+                    TargetBlockRetrievalPayload::TargetBlockId(target_block_id),
+                    num_blocks,
+                )
+            },
+            Some(window_size) => {
+                // For execution, we need the window starting from the highest_commit_cert + 1.
+                let highest_commit_cert_round =
+                    highest_commit_cert.ledger_info().ledger_info().round();
+                let target_round = max(
+                    1,
+                    min(
+                        highest_commit_cert_round,
+                        highest_commit_cert_round.saturating_sub(window_size),
+                    ),
+                );
+                let num_blocks = highest_quorum_cert.certified_block().round() - target_round + 1;
```

Comment:
it's always being a challenge to reconcile num_blocks with number of rounds, they don't necessarily align